### PR TITLE
BYTE_ARRAY Support: Add support to read/write byte[] data using raw index.

### PR DESCRIPTION
**Pull request by mayankshriv**
This is one of many PR's for BYTE_ARRAY data type support.
- Enhanced fixed and variable length Chunk based readers/writers to
  support byte-arrays.
- Enhanced SingleColumnSingleValueReader to support reading data as
  byte-arrays.
- Added unit tests to cover the new functionality.

### CODE REVIEW

File: pinot-core/src/main/java/com/linkedin/pinot/core/io/reader/SingleColumnSingleValueReader.java [hunk 1]
```diff
@@ -83,5 +83,13 @@
    */
   byte[] getBytes(int row);
 
+  /**
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/com/linkedin/pinot/core/io/reader/SingleColumnSingleValueReader.java]: Please add a description for the method

File: pinot-core/src/main/java/com/linkedin/pinot/core/data/readers/PinotSegmentColumnReader.java [hunk 2]
```diff
@@ -95,6 +95,16 @@ Object readString(int docId) {
     }
   }
 
+  Object readBytes(int docId) {
```
[the remark this answers is not in the document]
Comment (mayankshriv) [reply pinot-core/src/main/java/com/linkedin/pinot/core/data/readers/PinotSegmentColumnReader.java]: For symmetry with other api's in this class. This code is not in any critical path, and only used for reading pinot segment format (for conversions etc), so will keep it the same way.

File: pinot-core/src/main/java/com/linkedin/pinot/core/io/reader/impl/v1/FixedByteChunkSingleValueReader.java [hunk 3]
```diff
@@ -108,6 +111,28 @@ public double getDouble(int row, ChunkReaderContext context) {
...
+      byte[] bytes = _reusableBytes.get();
+      getRawData().copyTo(row * _lengthOfLongestEntry, bytes, 0, _lengthOfLongestEntry);
+      return bytes;
+    } else {
+      throw new UnsupportedOperationException("Read without context not supported for compressed data.");
+    }
+  }
+
+  @Override
+  public byte[] getBytes(int row, ChunkReaderContext context) {
```
[the remark this answers is not in the document]
Comment (mayankshriv) [reply pinot-core/src/main/java/com/linkedin/pinot/core/io/reader/impl/v1/FixedByteChunkSingleValueReader.java]: The returned byte[] is of exact size.

File: pinot-core/src/main/java/com/linkedin/pinot/core/io/reader/impl/v1/VarByteChunkSingleValueReader.java [hunk 4]
```diff
@@ -58,29 +56,65 @@ public String getString(int row, ChunkReaderContext context) {
...
     return new String(bytes, 0, length, UTF_8);
   }
 
+  @Override
+  public byte[] getBytes(int row) {
+    return getBytes(row, createContext());
+  }
+
+  @Override
+  public byte[] getBytes(int row, ChunkReaderContext context) {
```
[the remark this answers is not in the document]
Comment (mayankshriv) [reply pinot-core/src/main/java/com/linkedin/pinot/core/io/reader/impl/v1/VarByteChunkSingleValueReader.java]: Yes we do, the context holds the uncompressed chunk buffer, so we avoid reading and uncompressing of chunk buffers.

File: pinot-core/src/test/java/com/linkedin/pinot/index/readerwriter/VarByteChunkSingleValueReaderWriteTest.java [hunk 5]
```diff
@@ -84,9 +86,11 @@ public void test()
     VarByteChunkSingleValueReader reader = new VarByteChunkSingleValueReader(pinotDataBuffer);
     ChunkReaderContext context = reader.createContext();
 
-    for (int i = 0; i < NUM_STRINGS; i++) {
+    for (int i = 0; i < NUM_ENTRIES; i += 2) {
       String actual = reader.getString(i, context);
```
[the remark this answers is not in the document]
Comment (mayankshriv) [reply pinot-core/src/test/java/com/linkedin/pinot/index/readerwriter/VarByteChunkSingleValueReaderWriteTest.java]: Yes there is, see line 93.

**Pull request merged**